src/table/internal.tsx

```diff
@@ -152,12 +154,17 @@ const InternalTable = React.forwardRef(
     const hasFooter = !!footer;
 
     const visibleColumnsWithSelection = useMemo(() => {
-      const columnIds = visibleColumnDefinitions.map((it, index) => it.id ?? index.toString());
-      return hasSelection ? [selectionColumnId.toString(), ...columnIds] : columnIds ?? [];
+      const visible = visibleColumnDefinitions.map((c, i) => ({ ...c, id: getColumnKey(c, i).toString() }));
+      return hasSelection ? [{ id: selectionColumnId, width: SELECTION_CELL_WIDTH }, ...visible] : visible;
     }, [visibleColumnDefinitions, hasSelection]);
 
+    const visibleColumnIdsWithSelection = useMemo(
```

Comment:
I don't think it has much utility - the reason I wrapped the new variable in useMemo is to guarantee everything works the same after the refactor. I will play with it and remove both memos if that would appear safe.

Reply:
Apparently the memoization is needed for sticky columns. The table may re-render when scrolling upon reaching edges and also during the first load while the visibleColumns stay constant. Whenever visibleColumns change that triggers an effect inside the sticky columns util. I wouldn't touch this for now.

Reply:
The good practice is to memoise the stuff next to the effect call site. So it is clearly visible what it is memoized for. 

The memoization here depends on customer provided values, and this may lead to bugs, because we do not have an official requirement about reference equality of the columns and preferences.

I recommend fixing it either in this CR or in the very next one